// @keenoho/vite-plugin-compress-file/index.js
import fs from 'fs';
import path from 'path';
import zlib from 'zlib';

export function compressFile(subFilePath) {
  const readStream = fs.createReadStream(subFilePath);
  const writeStream = fs.createWriteStream(`${subFilePath}.gz`);
  const gzip = zlib.createGzip();
  readStream.pipe(gzip).pipe(writeStream);
}

export default function compressFilePlugin(options) {
  let config;

  return {
    name: 'vite-plugin-compress-file',

    configResolved(resolvedConfig) {
      config = resolvedConfig;
    },

    closeBundle(err) {
      if (err) {
        console.log(err);
        return;
      }

      let { root, outDir, exclude = ['.DS_Store', '.gitkeep', /^\./] } = options || {};
      root = root || config?.inlineConfig?.root || config?.root || process.cwd();
      outDir = config?.inlineConfig?.build.outDir || config?.build.outDir || 'dist';
      const distDir = path.resolve(root, outDir);

      function loopFindFile(parentPath) {
        if (!fs.existsSync(parentPath)) {
          return;
        }
        const parentStat = fs.statSync(parentPath);
        if (!parentStat.isDirectory()) {
          return;
        }

        for (let subFile of fs.readdirSync(parentPath)) {
          const subFilePath = path.resolve(parentPath, subFile);
          if (!fs.existsSync(subFilePath)) {
            continue;
          }
          if (exclude && exclude.length) {
            let isExclude = false;
            for (const e of exclude) {
              if (typeof e === 'string' && e == subFile) {
                isExclude = true;
                break;
              } else if (typeof e === 'function' && e(subFile)) {
                isExclude = true;
                break;
              } else if (e instanceof RegExp && e.test(subFile)) {
                isExclude = true;
                break;
              }
            }
            if (isExclude) {
              continue;
            }
          }
          const stat = fs.statSync(subFilePath);
          if (stat.isFile()) {
            compressFile(subFilePath);
          } else if (stat.isDirectory()) {
            loopFindFile(subFilePath);
          }
        }
      }

      loopFindFile(distDir);
    },
  };
}
